kernel/rros/observable.rs

```diff
@@ -0,0 +1,968 @@
+}
+
+impl GetLinks for RrosObserver {
+    type EntryType = RrosObserver;
+    fn get_links(data: &Self::EntryType) -> &Links<Self::EntryType> {
+        &data.next
+    }
+}
+
+fn add_subscription(
+    observable: &mut RrosObservable,
+    backlog_count: u32,
+    op_flags: i32,
+) -> Result<i32> {
+    let mut sbr = dovetail::dovetail_current_state().subscriber() as *mut RrosSubscriber;
+    if backlog_count == 0 {
+        return Err(kernel::Error::EINVAL);
+    }
+    if op_flags & (!RROS_NOTIFY_MASK) != 0 {
+        return Err(kernel::Error::EINVAL);
+    }
+    if sbr == 0 as *mut RrosSubscriber {
+        match Box::try_new(RrosSubscriber::new_and_init()) {
+            Ok(s) => sbr = Box::into_raw(s),
+            Err(_) => return Err(kernel::Error::ENOMEM),
+        }
+
+        dovetail::dovetail_current_state().set_subscriber(sbr);
+    }
+
+    // rros_get_element();
+
+    let backlog_count = backlog_count as usize;
+    let backlog_size = backlog_count * size_of::<RrosNotificationRecord>();
+
+    let mut observer = RrosObserver::new();
+    let fundle = observable.element.borrow().fundle;
+    observer.fundle = fundle;
+    observer.backlog_size = backlog_size;
+    observer.flags = op_flags;
+    observer.refs = 1;
+    if op_flags & RROS_NOTIFY_ONCHANGE != 0 {
+        observer.flags |= RROS_NOTIFY_INITIAL;
+    }
+    // Build the free list of notification records
+    // I don't use the `backlog` field, but alloc RrosNotificationRecord in heap and push it in free_list straightly
+    for _ in 0..backlog_count {
+        let nf = match Box::try_new(RrosNotificationRecord::new()) {
+            Ok(n) => n,
+            Err(_) => {
+                return Err(kernel::Error::ENOMEM);
+            }
+        };
+        observer.free_list.push_back(nf);
+    }
+
+    let observer = match Arc::try_new(observer) {
+        Ok(o) => o,
+        Err(_) => {
+            // TODO:
+            // evl_put_element(&observable->element);
+            return Err(kernel::Error::ENOMEM);
+        }
+    };
+
+    let subscriber = unsafe { &mut (*sbr) };
+    let flags = subscriber.subscriptions.irq_lock_noguard();
+    let ret = unsafe {
+        (*(subscriber.subscriptions.locked_data().get())).try_insert(fundle, observer.clone())
+    };
+    subscriber.subscriptions.irq_unlock_noguard(flags);
+
+    match ret {
+        Ok(None) => {
+            // everything is right
+        }
+        Ok(Some(_)) => {
+            // allocate a Node, but RBTree has a node with the same key
+            // node will be free automatically
+            return Err(kernel::Error::EEXIST);
+        }
+        Err(info) => {
+            // can't allocate a Node
+            // observer will be free automatically
+            return Err(info);
+        }
+    }
+
+    let flags = observable.lock.raw_spin_lock_irqsave();
+    observable.observers.push_back(observer);
+
+    observable.oob_wait.lock.raw_spin_lock();
+    observable.writable_observers += 1;
```

Comment:
> We need the observer and writability tracking to be atomically updated, so nested locking is required. Observers are linked to the observable's list by order of subscription (FIFO) in case users make such assumption for master mode (which undergoes round-robin).  This notation is taken from the same section of evl. I definitely should include this comment in the code. However, I think that in future refactoring, we might consider using Arc to replace this implementation.